astronomy/time_scales_body.hpp

```diff
@@ -570,6 +571,44 @@ inline Instant Parse北斗Time(std::string const& s) {
   return operator""_北斗(s.c_str(), s.size());
 }
 
+inline Date TTDay(Instant const& t) {
+  std::int64_t mjd = std::floor((t - "MJD0"_TT) / Day);
+  // We want operations rounded toward negative infinity, but we also don’t
+  // want to fiddle with rounding modes.  The product and sum here should be
+  // exact for all reasonable times.
+  if (mjd * Day + "MJD0"_TT > t) {
+    --mjd;
+  }
+  int digits = 0;
+  for (std::int64_t mjd_most_significant_digits = mjd;
+       mjd_most_significant_digits != 0;
+       ++digits) {
+    mjd_most_significant_digits /= 10;
+  }
+  return JulianDate::MJD(mjd, digits, /*fractional_digit_count=*/0).CalendarDay();
+}
+
+inline DateTime TTSecond(Instant const& t) {
+  auto const date = TTDay(t);
+  Instant const beginning_of_day = DateTimeAsTT(DateTime::BeginningOfDay(date));
+  // Close to J2000, Sterbenz’s lemma can fail to apply to this subtraction.  We
+  // compute it exactly and round by hand.
```

Comment:
This is an unfortunate property.  I wish we had taken, e.g., the origin of MJD as our zero.  (Nobody flies rockets in 1858.)

Reply:
JD0 would be the standard approach if the origin is really annoying, but I don’t think this really is a problem except for this kind of correct rounding or round-tripping trickery, and having the origin in the middle of the timespan we care about means we have better resolution, which seems like it should be more broadly useful.